src/PowerPing/Ping.cs

```diff
@@ -415,38 +399,54 @@ private PingResults SendICMP(PingAttributes attrs)
                     }
 
                     ICMP response;
-                    TimeSpan replyTime;
+                    EndPoint responseEP = iep;
+                    TimeSpan replyTime = TimeSpan.Zero;
                     do {
+                        // Cancel if requested
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        // Set receive timeout
+                        int remainingTimeout = (int)Math.Ceiling(attrs.Timeout - replyTime.TotalMilliseconds);
+                        if (remainingTimeout <= 0) {
+                            throw new SocketException();
+                        }
+                        sock.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReceiveTimeout, Math.Min(remainingTimeout, 250));
```

Comment:
So each iteration of this do..while we are decrementing the overall timeout by 250ms, yet we are setting the socket timeout each time. Doesn't seem right, because most iterations we are going to be unnecessarily calling setSocketOption to set the same value (250). Even if we have to use conditional logic to work around that I think its better and faster than using method or object calls too much.

Also like I mentioned, maybe I'm alergic to System.Net.Math but couldn't we just inline it here to use remainingTime if its less than 250?

` sock.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReceiveTimeout, remainingTimeout > 250 ? 250 : remainingTimeout);`

Like I bet the Math library is pretty fast but in general I really want to avoid unnecessary overhead if at all possible in this part of the code and keep this core sending/recieving loop as object light as it can be.

Reply:
Yeah I should probably add logic to set the timeout only if it's changing. I did it that was just for simplicity as I didn't think it would affect performance too much.

As for inlining Math.Min, it's definitely not necessary. If you look at the code on Reference Source, it's just "return (val1<=val2)?val1:val2;", and I verified on SharpLab.io by checking the JIT Asm output that the JIT automatically inlines it.